[5.2] Vote : add aggregateRating in article's schemaorg information

conseilgouz opened the pull request:
( see discussion https://github.com/joomla/joomla-cms/discussions/42920)

### Summary of Changes
When vote plugin is enabled, Rich results are created but are rejected by Google Rich Results tool.


### Testing Instructions
Enable Vote plugin.
Enable Voting in Articles component configuration.
Vote for one or more articles
Set different schemaorg type for your articles
Check your articles using https://search.google.com/test/rich-results

### Actual result BEFORE applying this Pull Request
Google Rich Results check gives an error 
Review snippets shows 2 errors : 
![rich_snippet_error](https://github.com/joomla/joomla-cms/assets/19435246/facdb827-4df2-4be7-bcc8-f148363fa689)

### Expected result AFTER applying this Pull Request
Google Rich Results check is OK
![rich_snippet_ok](https://github.com/joomla/joomla-cms/assets/19435246/ef3fc2d6-c263-4c8c-a4ec-c46616e1a9c5)

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [X] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [X] No documentation changes for manual.joomla.org needed

Commit:
[5.1]Vote : aggregateRating error in Google Rich Results

ManuelVoileux commented:
Thanks for raising hand.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42933">issues.joomla.org/tracker/joomla-cms/42933</a>.</sub>

conseilgouz commented:
This fix works with Joomla! 5.0 but, as there won't be any more 5.0 release, I put it in 5.1 branch.

ManuelVoileux commented:
I have experienced the same issue.
I have activated the Vote plugin, and I have used vote on an article.
When I check with the Google validation tool I get a critical error. itemReviewed field missing.
I guess that the itemtype, itemscope, should come from the input of the schema tab, and the output of the rating.php file should be merged in the schema plugin output.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42933">issues.joomla.org/tracker/joomla-cms/42933</a>.</sub>

ManuelVoileux commented:
AFTER applying this Pull Request I got green light from Google
![Screenshot 2024-03-01 at 13-30-10](https://github.com/joomla/joomla-cms/assets/117023571/367286fd-c58a-4114-b456-8ab1572ca123)

sandewt commented:
This issue also concerns Joomla 4.

richard67 commented:
> AFTER applying this Pull Request I got green light from Google

@ManuelVoileux Does that mean a successful test? If so, please go to the issue tracker here https://issues.joomla.org/tracker/joomla-cms/42933 and mark your test result by using the blue "Test this" button at the top left corner, then selecting the appropriate test result (success) and submit.

Every pull request needs 2 successful human tests, and it needs to submit the test result with the issue tracker so the tests are properly counted.

Thanks in advance.

richard67 commented:
> This issue also concerns Joomla 4.

@conseilgouz If this is right, it would need to make this pull request for the 4.4-dev branch. Later maintainers will merge it up into the 5.1-dev branch when the 4.4-dev PR will have been tested and merged.

brianteeman commented:
This solution is correct for J4 which uses itemprop etc
This solution is not correct for J5 which used jsonld

ManuelVoileux commented:
I have tested this item :white_check_mark: successfully on ceb86b192a7bd305bb9fd788ea91f6a453571cc0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42933">issues.joomla.org/tracker/joomla-cms/42933</a>.</sub>

Fedik commented:
Not that easy. 

The page does not contain Product, how it can have the product rating then?
I mean having an empty Product is invalid also.

Just try https://search.google.com/test/rich-results?hl=en
```
<div itemscope="itemscope" itemtype="https://schema.org/Product"> 
<p class="visually-hidden" itemprop="aggregateRating" itemscope itemtype="https://schema.org/AggregateRating">
   <meta itemprop="ratingCount" content="5">
   <meta itemprop="worstRating" content="1">
   <span itemprop="ratingValue">5</span>
   <span itemprop="bestRating">5</span>
</p>
</div>
```

It will ask you for full product information.

Fedik commented:
A better fix would be to remove all HTML schema from layout `plugins/content/vote/tmpl/rating.php`
And add it with schema plugin, as part of the content schema.

ManuelVoileux commented:
> Not that easy.
> The page does not contain Product, how it can have the product rating then? I mean having an empty Product is invalid also.
> Just try https://search.google.com/test/rich-results?hl=en
> 
> ```
> <div itemscope="itemscope" itemtype="https://schema.org/Product"> 
> <p class="visually-hidden" itemprop="aggregateRating" itemscope itemtype="https://schema.org/AggregateRating">
>    <meta itemprop="ratingCount" content="5">
>    <meta itemprop="worstRating" content="1">
>    <span itemprop="ratingValue">5</span>
>    <span itemprop="bestRating">5</span>
> </p>
> </div>
> ```
> 
> It will ask you for full product information.

If the Schema tab is filled in,  the itemscope, itemtype should be taken from the schema type, and the rating result should be merged in the schema output.

conseilgouz commented:
> The page does not contain Product, how it can have the product rating then?

I had a doubt about Product type. This idea comes from JED pages : if you take a look at https://extensions.joomla.org/extension/cg-isotope/ source code, it's done this way and Google search gives me nice stars.

conseilgouz commented:
![google_search](https://github.com/joomla/joomla-cms/assets/19435246/59f973b9-e94f-470c-894d-571eaf4645f7)

conseilgouz commented:
> > This issue also concerns Joomla 4.
> 
> @conseilgouz If this is right, it would need to make this pull request for the 4.4-dev branch. Later maintainers will merge it up into the 5.1-dev branch when the 4.4-dev PR will have been tested and merged.

Same code, same issue in Joomla 4.4.

Is 4.4-dev PR a new PR ?

Commit:
Update rating.php

Missing item name

conseilgouz commented:
closed, move to 4.4-dev branch #42934

sandewt commented:
> closed, move to 4.4-dev branch https://github.com/joomla/joomla-cms/pull/42934

You could have changed the branche here to move to 4.4

Fedik commented:
> I had a doubt about Product type. This idea comes from JED pages : if you take a look at 
> https://extensions.joomla.org/extension/cg-isotope/ source code, it's done this way and Google search gives me nice stars.

No, that not correct example.
That whole page is about Product, and there it is correct. 
Hovewer what you trying here, is incorect, sorry. Because here no product exists.

conseilgouz commented:
> No, that not correct example.
> That whole page is about Product, and there it is correct.
> Hovewer what you trying here, is incorect, sorry. Because here no product exists.

Product is a generic type, it could be anything : https://schema.org/Product

In JED page, I did not find any other schema.org line other than reviews and  aggregateRating, and Google seems to be happy with this.

conseilgouz commented:
> You could have changed the branche here to move to 4.4

I tried but I broke it (I'm not a great github user).

sandewt commented:
> Product is a generic type, it could be anything : https://schema.org/Product

It may be more applicable to use ?

`<div itemprop="review" itemscope itemtype="https://schema.org/Review">`

instead of:

 `<div itemprop="review" itemscope itemtype="https://schema.org/Product">`

see: https://schema.org/reviewCount
see: https://schema.org/Rating

See examples

Commit:
[5.1]  Add aggregateRating

Commit:
Merge branch '5.1-dev' into 5.1-dev

brianteeman commented on administrator/language/en-GB/plg_schemaorg_aggregaterating.ini:
```diff
@@ -0,0 +1,8 @@
+; Joomla! Project
+; (C) 2023 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
; (C) 2024 Open Source Matters, Inc. <https://www.joomla.org>
```

conseilgouz commented:
Just to make sure I'm going the correct way, I created a new plugin plugins/schemaorg/aggregaterating.

It adds aggregaterating infos to Recipe type in Json+ld format. 
For other types, it adds an aggregateRating lines with product type.

If it sounds correct, I'll continue adding AggregateRating to other schemaorg types.

brianteeman commented on administrator/language/en-GB/plg_schemaorg_aggregaterating.sys.ini:
```diff
@@ -0,0 +1,7 @@
+; Joomla! Project
+; (C) 2023 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
; (C) 2024 Open Source Matters, Inc. <https://www.joomla.org>
```

brianteeman commented on libraries/src/Schemaorg/SchemaorgPrepareProductAggregateRating.php:
```diff
@@ -0,0 +1,46 @@
+/**
+ * Prepare Product AggregateRating to be valid for JSON-LD output
+ *
+ * @since  5.0.0
```
```suggestion
 * @since  __DEPLOY_VERSION__
```

brianteeman commented on libraries/src/Schemaorg/SchemaorgPrepareProductAggregateRating.php:
```diff
@@ -0,0 +1,46 @@
+     *
+     * @return  ?string
+     *
+     * @since   5.0.0
```
```suggestion
 * @since  __DEPLOY_VERSION__
```

Commit:
Update libraries/src/Schemaorg/SchemaorgPrepareProductAggregateRating.php

Co-authored-by: Brian Teeman <brian@teeman.net>

brianteeman commented on libraries/src/Schemaorg/SchemaorgPrepareRecipeAggregateRating.php:
```diff
@@ -0,0 +1,46 @@
+/**
+ * Prepare Recipe AggregateRating to be valid for JSON-LD output
+ *
+ * @since  5.0.0
```
```suggestion
 * @since  __DEPLOY_VERSION__
```

brianteeman commented on libraries/src/Schemaorg/SchemaorgPrepareRecipeAggregateRating.php:
```diff
@@ -0,0 +1,46 @@
+     *
+     * @return  ?string
+     *
+     * @since   5.0.0
```
```suggestion
 * @since  __DEPLOY_VERSION__
```

Commit:
Update libraries/src/Schemaorg/SchemaorgPrepareProductAggregateRating.php

Co-authored-by: Brian Teeman <brian@teeman.net>

brianteeman commented on libraries/src/Schemaorg/SchemaorgPrepareRecipeAggregateRating.php:
```diff
@@ -0,0 +1,46 @@
+trait SchemaorgPrepareRecipeAggregateRating
+{
+    /**
+     * Prepare duration
```
duration ?? is this a copy paste error

conseilgouz replied:
yes

Commit:
Update administrator/language/en-GB/plg_schemaorg_aggregaterating.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

Commit:
Update libraries/src/Schemaorg/SchemaorgPrepareRecipeAggregateRating.php

Co-authored-by: Brian Teeman <brian@teeman.net>

Commit:
Update libraries/src/Schemaorg/SchemaorgPrepareRecipeAggregateRating.php

Co-authored-by: Brian Teeman <brian@teeman.net>

brianteeman commented on plugins/schemaorg/aggregaterating/aggregaterating.xml:
```diff
@@ -0,0 +1,21 @@
+<extension type="plugin" group="schemaorg" method="upgrade">
+	<name>plg_schemaorg_aggregaterating</name>
+	<author>Joomla! Project</author>
+	<creationDate>2023-07</creationDate>
```
```suggestion
	<creationDate>2024-03</creationDate>
```

brianteeman commented on plugins/schemaorg/aggregaterating/aggregaterating.xml:
```diff
@@ -0,0 +1,21 @@
+	<name>plg_schemaorg_aggregaterating</name>
+	<author>Joomla! Project</author>
+	<creationDate>2023-07</creationDate>
+	<copyright>(C) 2023 Open Source Matters, Inc.</copyright>
```
```suggestion
	<copyright>(C) 2024 Open Source Matters, Inc.</copyright>
```

brianteeman commented on plugins/schemaorg/aggregaterating/aggregaterating.xml:
```diff
@@ -0,0 +1,21 @@
+	<license>GNU General Public License version 2 or later; see LICENSE.txt</license>
+	<authorEmail>admin@joomla.org</authorEmail>
+	<authorUrl>www.joomla.org</authorUrl>
+	<version>5.0.0</version>
```
```suggestion
	<version>__DEPLOY_VERSION__</version>
```

brianteeman commented on plugins/schemaorg/aggregaterating/services/provider.php:
```diff
@@ -0,0 +1,47 @@
+ * @package     Joomla.Plugin
+ * @subpackage  Schemaorg.AggregateRating
+ *
+ * @copyright   (C) 2023 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
 * @copyright   (C) 2024 Open Source Matters, Inc. <https://www.joomla.org>
```

brianteeman commented on plugins/schemaorg/aggregaterating/services/provider.php:
```diff
@@ -0,0 +1,47 @@
+     *
+     * @return  void
+     *
+     * @since   5.0.0
```
```suggestion
 * @since  __DEPLOY_VERSION__
```

brianteeman commented on plugins/schemaorg/aggregaterating/src/Extension/AggregateRating.php:
```diff
@@ -0,0 +1,101 @@
+ * @package     Joomla.Plugin
+ * @subpackage  Schemaorg.AggregateRating
+ *
+ * @copyright   (C) 2023 Open Source Matters, Inc. <https://www.joomla.org>
```
```suggestion
 * @copyright   (C) 2024 Open Source Matters, Inc. <https://www.joomla.org>
```

brianteeman commented on plugins/schemaorg/aggregaterating/src/Extension/AggregateRating.php:
```diff
@@ -0,0 +1,101 @@
+/**
+ * Schemaorg Plugin
+ *
+ * @since  5.0.0
```
```suggestion
 * @since  __DEPLOY_VERSION__
```

brianteeman commented on plugins/schemaorg/aggregaterating/src/Extension/AggregateRating.php:
```diff
@@ -0,0 +1,101 @@
+     * Load the language file on instantiation.
+     *
+     * @var    boolean
+     * @since  5.0.0
```
```suggestion
 * @since  __DEPLOY_VERSION__
```

brianteeman commented on plugins/schemaorg/aggregaterating/src/Extension/AggregateRating.php:
```diff
@@ -0,0 +1,101 @@
+     * The name of the schema form
+     *
+     * @var   string
+     * @since 5.0.0
```
```suggestion
 * @since  __DEPLOY_VERSION__
```

brianteeman replied:
```suggestion
     * @since  __DEPLOY_VERSION__
```

Commit:
Update plugins/schemaorg/aggregaterating/aggregaterating.xml

Co-authored-by: Brian Teeman <brian@teeman.net>

Commit:
Update plugins/schemaorg/aggregaterating/services/provider.php

Co-authored-by: Brian Teeman <brian@teeman.net>

Commit:
Update plugins/schemaorg/aggregaterating/aggregaterating.xml

Co-authored-by: Brian Teeman <brian@teeman.net>

Commit:
Update plugins/schemaorg/aggregaterating/services/provider.php

Co-authored-by: Brian Teeman <brian@teeman.net>

brianteeman commented on plugins/schemaorg/aggregaterating/src/Extension/AggregateRating.php:
```diff
@@ -0,0 +1,101 @@
+     *
+     * @return  array
+     *
+     * @since   5.0.0
```
```suggestion
     * @since  __DEPLOY_VERSION__
```

Commit:
Update plugins/schemaorg/aggregaterating/src/Extension/AggregateRating.php

Co-authored-by: Brian Teeman <brian@teeman.net>

Commit:
Update plugins/schemaorg/aggregaterating/src/Extension/AggregateRating.php

Co-authored-by: Brian Teeman <brian@teeman.net>

brianteeman commented on plugins/schemaorg/aggregaterating/src/Extension/AggregateRating.php:
```diff
@@ -0,0 +1,101 @@
+     *
+     * @return  void
+     *
+     * @since   5.0.0
```
```suggestion
     * @since  __DEPLOY_VERSION__
```

Commit:
Update plugins/schemaorg/aggregaterating/src/Extension/AggregateRating.php

Co-authored-by: Brian Teeman <brian@teeman.net>

Commit:
Update plugins/schemaorg/aggregaterating/aggregaterating.xml

Co-authored-by: Brian Teeman <brian@teeman.net>

Commit:
Update plugins/schemaorg/aggregaterating/src/Extension/AggregateRating.php

Co-authored-by: Brian Teeman <brian@teeman.net>

Commit:
Update plugins/schemaorg/aggregaterating/src/Extension/AggregateRating.php

Co-authored-by: Brian Teeman <brian@teeman.net>

Commit:
Update plugins/schemaorg/aggregaterating/src/Extension/AggregateRating.php

Co-authored-by: Brian Teeman <brian@teeman.net>

brianteeman commented:
Please see  #42699 for an example of all the other changes that you need to make to add this new plugin

Commit:
Update SchemaorgPrepareProductAggregateRating.php

Commit:
Update SchemaorgPrepareRecipeAggregateRating.php

bembelimen commented:
You should not add a new plugin, the existing plugin should do what @Fedik said here: https://github.com/joomla/joomla-cms/pull/42934#issuecomment-1975127605

conseilgouz commented:
> You should not add a new plugin, the existing plugin should do what @Fedik said here: [#42934 (comment)](https://github.com/joomla/joomla-cms/pull/42934#issuecomment-1975127605)

sounds easier

Commit:
5.1 : AggregateRating

Commit:
5.1 : aggregaterating

Commit:
Update rating.php

Commit:
5.1 aggregateRating

remove tabs

Commit:
Update Vote.php

Commit:
Merge branch '5.1-dev' into 5.1-dev

Commit:
5.1 : aggregatrating

add new lines at end of files

Commit:
Merge branch '5.1-dev' of https://github.com/conseilgouz/joomla-cms into 5.1-dev

Commit:
5.1 : aggregaterating

Commit:
5.1 : aggregaterating

Commit:
Merge branch '5.1-dev' into 5.1-dev

Commit:
5.1 : aggregaterating

remove extra spaces

Commit:
Merge branch '5.1-dev' of https://github.com/conseilgouz/joomla-cms into 5.1-dev

Commit:
Update Vote.php

Commit:
5.1 : aggregaterating

still some extra spaces/tabs to remove

Commit:
Update SchemaorgPrepareRecipeAggregateRating.php

Quy commented on plugins/content/vote/tmpl/rating.php:
```diff
@@ -79,11 +79,6 @@
 <div class="content_rating" role="img" aria-label="<?php echo Text::sprintf('PLG_VOTE_STAR_RATING', $rating); ?>">
     <?php if ($rcount) : ?>
-        <p class="visually-hidden" itemprop="aggregateRating" itemscope itemtype="https://schema.org/AggregateRating">
-            <?php echo Text::sprintf('PLG_VOTE_USER_RATING', '<span itemprop="ratingValue">' . $rating . '</span>', '<span itemprop="bestRating">5</span>'); ?>
```
Deprecate language string.

conseilgouz replied:
it cleaned up too much, thank you for telling me.

Commit:
Update rating.php

Quy commented on libraries/src/Schemaorg/SchemaorgPrepareProductAggregateRating.php:
```diff
@@ -0,0 +1,47 @@
+     */
+    protected function prepareProductAggregateRating($context)
+    {
+        [$extension, $view, $id] = explode('.', $context);
+        if ($view == 'article') {
+            $baseId   = Uri::root() . '#/schema/';
+            $schemaId = $baseId . str_replace('.', '/', $context);
+
+            $component = $this->getApplication()->bootComponent('com_content')->getMVCFactory();
+            $model     = $component->createModel('Article', 'Site');
+            $article   = $model->getItem($id);
+            return ['@isPartOf' => ['@id' => $schemaId,'@type' => 'Product','name' => $article->title,'aggregateRating' => ['@type' => "AggregateRating","ratingCount" => (string)$article->rating_count,"ratingValue" => (string)$article->rating]]];
+        }
+        return false;
```
```suggestion
        [$extension, $view, $id] = explode('.', $context);

        if ($view === 'article') {
            $baseId   = Uri::root() . '#/schema/';
            $schemaId = $baseId . str_replace('.', '/', $context);

            $component = $this->getApplication()->bootComponent('com_content')->getMVCFactory();
            $model     = $component->createModel('Article', 'Site');
            $article   = $model->getItem($id);

            return ['@isPartOf' => ['@id' => $schemaId, '@type' => 'Product', 'name' => $article->title, 'aggregateRating' => ['@type' => 'AggregateRating', 'ratingCount' => (string) $article->rating_count, 'ratingValue' => (string) $article->rating]]];
        }

        return false;
```

Quy replied:
Please apply the same coding style in the other file. Thanks.

Quy commented on plugins/content/vote/src/Extension/Vote.php:
```diff
@@ -123,4 +127,39 @@ private function displayVotingData($context, &$row, &$params, $page)
+
+        if (!$need_vote) {
+            return;
+        }
+        foreach ($graph as $key => &$entry) {
+            if (!isset($entry['@type'])) {
+                continue;
+            }
+            if ($entry['@type'] == 'Recipe') {
+                $rating = $this->prepareRecipeAggregateRating($ontext);
+                continue;
+            }
+            $rating = $this->prepareProductAggregateRating($context);
+        }
```
```suggestion
        }

        foreach ($graph as $key => &$entry) {
            if (!isset($entry['@type'])) {
                continue;
            }

            if ($entry['@type'] === 'Recipe') {
                $rating = $this->prepareRecipeAggregateRating($ontext);
                continue;
            }

            $rating = $this->prepareProductAggregateRating($context);
        }
```

Commit:
Update libraries/src/Schemaorg/SchemaorgPrepareProductAggregateRating.php

Co-authored-by: Quy <quy@nomonkeybiz.com>

Commit:
Update plugins/content/vote/src/Extension/Vote.php

Co-authored-by: Quy <quy@nomonkeybiz.com>

Commit:
5.1 : aggregaterating

Drone error

Commit:
Merge branch '5.1-dev' of https://github.com/conseilgouz/joomla-cms into 5.1-dev

Commit:
5.1 : aggregaterating

add extra types

Commit:
Merge branch '5.1-dev' into 5.1-dev

sandewt commented on plugins/content/vote/src/Extension/Vote.php:
```diff
@@ -12,6 +12,9 @@
 
 use Joomla\CMS\Plugin\CMSPlugin;
 use Joomla\CMS\Plugin\PluginHelper;
+use Joomla\CMS\Uri\Uri;
```
alphabetically

conseilgouz replied:
Drone already told me, thank you

Commit:
5.1 :  aggregaterating

fix Drone errors

Commit:
Merge branch '5.1-dev' of https://github.com/conseilgouz/joomla-cms into 5.1-dev

Commit:
Update Vote.php

sandewt commented on libraries/src/Schemaorg/SchemaorgPrepareProductAggregateRating.php:
```diff
@@ -0,0 +1,50 @@
+// phpcs:disable PSR1.Files.SideEffects
+\defined('_JEXEC') or die;
+// phpcs:enable PSR1.Files.SideEffects
+use Joomla\CMS\Uri\Uri;
```
move to line 12 (up)

conseilgouz commented:
Just added Book/Event/Organization types.
After some more testings, JobPosting and Person types are not supposed to have aggregaterating.

Commit:
5.1 : aggregaterating

lost SchemaorgPrepareAggregateRating

Commit:
Update Vote.php

fix drone error

Commit:
5.1 : aggregaterating

Remove unnecessary code

Commit:
Update Vote.php

Commit:
Update Vote.php

Commit:
5.1 aggregaterating

Drone : fix comments

Commit:
5.1 : aggregaterating for Article

Add Article in types

conseilgouz commented:
Added schemaorg type Article

Commit:
Merge branch '5.1-dev' into 5.1-dev

Commit:
Merge branch '5.1-dev' into 5.1-dev

Fedik commented:
That looks better, thanks.
Couple notes:

Please remove the traits, they are not reusable in this implementation. Just keep that code withing the plugin itself.
Add full list of schemes which support `aggregateRating` https://schema.org/AggregateRating:
> Brand, CreativeWork, Event, Offer, Organization, Place, Product, Service

Fedik commented on plugins/content/vote/src/Extension/Vote.php:
```diff
@@ -123,4 +129,55 @@ private function displayVotingData($context, &$row, &$params, $page)
+    {
+        $graph = $schema->get('@graph');
+
+        $need_vote = PluginHelper::isEnabled('content', 'vote');
```
You can remove this. The plugin will not be executed when it is disabled.

conseilgouz replied:
done, test in progress

sandewt commented on plugins/content/vote/tmpl/rating.php:
```diff
@@ -79,11 +79,7 @@
-            <meta itemprop="ratingCount" content="<?php echo $rcount; ?>">
-            <meta itemprop="worstRating" content="1">
-        </p>
+        <?php echo Text::sprintf('PLG_VOTE_USER_RATING', $rating, 5); ?>
```
This text is not visible in the original
class="visually-hidden"
![Schermafbeelding 2024-03-04 131458](https://github.com/joomla/joomla-cms/assets/9271775/3dcab4f5-d91a-466f-a3d2-dfd867ad0670)

conseilgouz replied:
adding  back missing p + class

conseilgouz replied:
Test in progress

sandewt commented on plugins/content/vote/src/Extension/Vote.php:
```diff
@@ -123,4 +129,55 @@ private function displayVotingData($context, &$row, &$params, $page)
+            }
+        }
+
+        if ($rating) {
```
```suggestion
        if (isset($rating)) {
```

sandewt replied:
Warning
: Undefined variable $rating in
...\Joomla\plugins\content\vote\src\Extension\Vote.php
on line
178

conseilgouz replied:
$rating contains false if rating_count = 0
So it becomes 
`if (isset($rating) && $rating) { `

sandewt commented on libraries/src/Schemaorg/SchemaorgPrepareAggregateRating.php:
```diff
@@ -0,0 +1,52 @@
+    {
+        [$extension, $view, $id] = explode('.', $context);
+
+        if ($view == 'article') {
```
```suggestion
        if ($view === 'article') {
```

conseilgouz replied:
library removed, content merged in vote plugin.

Commit:
Update libraries/src/Schemaorg/SchemaorgPrepareAggregateRating.php

Co-authored-by: jsanders <j53.sanders@gmail.com>

Commit:
Update plugins/content/vote/src/Extension/Vote.php

Co-authored-by: jsanders <j53.sanders@gmail.com>

Commit:
Merge branch '5.1-dev' into 5.1-dev

Commit:
5.1 : aggregaterating

Merging aggregaterating functions into vote plugin
remove unnecessary trait
add full list of schemes supporting aggregaterating
add removed visually hidden class

Commit:
Merge branch '5.1-dev' of https://github.com/conseilgouz/joomla-cms into 5.1-dev

Commit:
Update Vote.php

Fix Drone error

Commit:
No vote

No vote : prepareProductAggregateRating return false if rating-count = 0

Commit:
Update Vote.php

Commit:
Update Vote.php

sandewt commented on plugins/content/vote/src/Extension/Vote.php:
```diff
@@ -123,4 +124,110 @@ private function displayVotingData($context, &$row, &$params, $page)
+            $model     = $component->createModel('Article', 'Site');
+            $article   = $model->getItem($id);
+            if ($article->rating_count > 0) {
+                return ['@isPartOf' => ['@id' => $schemaId, 'aggregateRating' => ['@type' => 'AggregateRating','ratingCount' => (string) $article->rating_count,'ratingValue' => (string) $article->rating]]];
```
Have you chosen to drop the worstRating and the bestRating from the schema?

conseilgouz replied:
worstRating and bestRating are optional and are not handle by Vote plugin nor content_rating table.

According to https://schema.org/AggregateRating , if absent, worstRating value is assumed to be 1 and bestRating is 5, useless fixed values that were in previous version of Vote plugin.

Of course, if you wish, I may include them again.

sandewt replied:
Thanks for your explanation. So these are default values. So I have no meaningful argument to include them.

sandewt commented on plugins/content/vote/src/Extension/Vote.php:
```diff
@@ -123,4 +124,110 @@ private function displayVotingData($context, &$row, &$params, $page)
+                continue;
+            }
+
+            switch ($entry['@type']) {
+                case 'Book':
+                case 'Brand':
+                case 'CreativeWork':
+                case 'Event':
+                case 'Offer':
+                case 'Organization':
+                case 'Place':
+                case 'Product':
+                case 'Recipe':
+                case 'Service':
+                    $rating = $this->prepareAggregateRating($context);
+                    break;
+                case 'Article':
+                case 'BlogPosting':
+                    $rating = $this->prepareProductAggregateRating($context);
+                    break;
```
> Brand, CreativeWork, Event, Offer, Organization, Place, Product, Service

@Fedik  , @conseilgouz 

This list differs from what plugins are available. Is that the intention? Or are they intended for extension builders?

![Schermafbeelding 2024-03-05 125141](https://github.com/joomla/joomla-cms/assets/9271775/5635b733-e358-4f39-b0af-e874ae72a27a)



![Schermafbeelding 2024-03-05 125735](https://github.com/joomla/joomla-cms/assets/9271775/d94fe9ed-462b-4d60-9eba-fb07fd74c64b)

[EDIT]

conseilgouz replied:
This weekend (which has been a long one), I saw a message (from Brian, I think, not sure) telling somebody to develop its own schemaorg plugin.
So, Fedik request is to make Vote plugin work in all cases

sandewt replied:
Tanks. Clearly. 
May I express my appreciation for your programming work here. Well done.

conseilgouz replied:
Thank you.
Let's wait for Fedik answer before closing this conversation, just to make sure I understood well his request's purpose.

Fedik replied:
Yes, it to make Vote plugin work in all cases. Example when someone will code own schema plugin.

conseilgouz replied:
thank you

Commit:
Merge branch '5.1-dev' into 5.1-dev

Commit:
Merge branch '5.1-dev' into 5.1-dev

Commit:
Merge branch '5.1-dev' into 5.1-dev

Quy commented:
@conseilgouz FYI: It is not necessary to keep updating the branch as it will also reset test results that maintainers have to restore. This will save you/maintainers time. Thank you.

exlemor commented:
I have tested this item :white_check_mark: successfully on 556e59ede0f5bbc61d4fd8f609b7276955312465<br /><br />I was able to replicate successfully the Expected results of Google Rich Results check is OK (and had the error prior to patch).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42933">issues.joomla.org/tracker/joomla-cms/42933</a>.</sub>

viocassel commented:
I have tested this item :white_check_mark: successfully on 556e59ede0f5bbc61d4fd8f609b7276955312465<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42933">issues.joomla.org/tracker/joomla-cms/42933</a>.</sub>

Quy commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42933">issues.joomla.org/tracker/joomla-cms/42933</a>.</sub>

Commit:
Merge branch 'joomla:5.1-dev' into 5.1-dev

Commit:
Update Update.php

Commit:
Update UpdateModel.php

Commit:
restore Update.php

Commit:
restore  UpdateModel.php

Commit:
Merge branch 'joomla:5.1-dev' into 5.1-dev

growunder commented:
I have tested this item :white_check_mark: successfully on 556e59ede0f5bbc61d4fd8f609b7276955312465<br /><br />Tested successfully. applied patch and fixed the issue

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42933">issues.joomla.org/tracker/joomla-cms/42933</a>.</sub>

cybersalt commented:
I have tested this item :white_check_mark: successfully on 556e59ede0f5bbc61d4fd8f609b7276955312465<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42933">issues.joomla.org/tracker/joomla-cms/42933</a>.</sub>

conseilgouz commented:
@cybersalt @growunder Thank you for your test. 
@exlemor and @viocassel already made "human tests", so now we have 4 positive tests.
Unfortunately, Fedik made this a new feature instead of a bug (why ?), so, even if is RTC, it won't be released with 5.1, may be in 5.2...
Pascal

pe7er committed:
Merge branch '5.2-dev' into 5.1-dev

pe7er committed:
Merge branch '5.2-dev' into 5.1-dev

pe7er committed:
Merge branch '5.2-dev' into 5.1-dev

QuyTon committed:
Merge branch '5.2-dev' into 5.1-dev

pe7er commented:
Thanks @conseilgouz !
